### DISCUSSION TITLE
Wire.h i2c slave  not working, STM32F103 is freezing..

### DISCUSSION THREAD

**Discussion by gopalreddylakki** [stm32duino/Arduino_Core_STM32#2279] [General]
![i2cs](https://github.com/stm32duino/Arduino_Core_STM32/assets/37047660/f1cd566d-4411-4899-a001-f423812cf301)
![Slave_STm32F103](https://github.com/stm32duino/Arduino_Core_STM32/assets/37047660/8b6bc0ee-2ce5-4085-a1a9-04945c65b357)
![Master_nano_read](https://github.com/stm32duino/Arduino_Core_STM32/assets/37047660/dc03a687-18be-4a28-b842-c3587bb3e491)

```C++
//slave
void requestEvent()
{
  Wire.write("hello\n");  // respond with message of 6 bytes
                          // as expected by master
}
```

form the Arduino required 15 bytes, in the salve only having 6 bytes.  in slave (stm32f103) after sending 6 bytes it's not the end transaction, it's in freezing. 

 but I check with another board manager [stm32duino](https://github.com/stm32duino) . it's working fine.

 can check it and give me the solution 

Slave program (STM32F103)

```C++
#include <Wire.h>

#define I2C_ADDR  0x09

void setup()
{
  Wire.begin(I2C_ADDR);         // join i2c bus with address #4
  Wire.onRequest(requestEvent); // register event
  Wire.onReceive(receiveEvent); // register event
  Serial.begin(115200);           // start serial for output
    pinMode(PC13,OUTPUT);
}

void loop()
{
  //empty loop
    digitalWrite(PC13,HIGH);
  delay(200);
  digitalWrite(PC13,LOW);
  delay(200);
}

// function that executes whenever data is received from master
// this function is registered as an event, see setup()
void receiveEvent(int howMany)
{
  while(1 < Wire.available()) // loop through all but the last
  {
    char c = Wire.read();     // receive byte as a character
    Serial.print(c);          // print the character
  }
 // int x = Wire.read();        // receive byte as an integer
  //Serial.println(x);          // print the integer
}

// function that executes whenever data is requested by master
// this function is registered as an event, see setup()
void requestEvent()
{
  Wire.write("hello\n");  // respond with message of 6 bytes
                          // as expected by master
}

```

master program (ardunio nano )

```C++
#include "Wire.h"
#define bl03_address 0x09

void setup()
{
 Wire.begin();
 Serial.begin(115200);
 
 
}
void loop()
{
  //getvoltage();
 //gettemperature();

	
  Wire.beginTransmission(bl03_address);
	Wire.write(0x20);
	Wire.endTransmission();
	Wire.requestFrom(bl03_address,15,true);
	char registerValue = Wire.read();
  char registerValue1 = Wire.read();
	Wire.endTransmission();
  Serial.println(registerValue,HEX);
 Serial.println(registerValue1,HEX);

 delay(500);
}

```

**Comment by gopalreddylakki** [comment]
[Ardunio_stm32](https://github.com/rogerclarkmelbourne/Arduino_STM32/tree/master)
has Wire_salve.h library with this library STM is working fine without freezing 
[STM32F1/libraries/WireSlave](https://github.com/rogerclarkmelbourne/Arduino_STM32/tree/master/STM32F1/libraries/WireSlave)
check it this library

**Reply by fpistm** [reply]
Hi @gopalreddylakki 
You are at the wrong place. This Github Organization is exclusively for the STM32core not the one from Roger Clarck.

**Reply by gopalreddylakki** [reply]
https://github.com/stm32duino/Arduino_Core_STM32

Wire.h is after sending the string it's freezing.  


in my application using SMBUS, in the communication, the master will ask for the address. multiple bytes it means more than the available string.

**Reply by fpistm** [reply]
Your issue is not clear, from my comprehension you used Roger clark core not the one provided here:

> but I check with another board manager [stm32duino](https://github.com/stm32duino) . it's working fine.

This suppose you used another core.

> [Ardunio_stm32](https://github.com/rogerclarkmelbourne/Arduino_STM32/tree/master)
> has Wire_salve.h library with this library STM is working fine without freezing

This suppose you found your issue.

**Reply by gopalreddylakki** [reply]
stm32core wire.h library having issues.

**Reply by fpistm** [reply]
Do you have I2C pullup on each i2C line, I've tested the default master/slave examples and it works as expected.

**Reply by gopalreddylakki** [reply]
Yes added the the pullup on SDA,SCL again sae issuse

Note
![LOGIC](https://github.com/stm32duino/Arduino_Core_STM32/assets/37047660/404bf38a-e50e-4a13-ad69-86df1b9ca898)
![STM32](https://github.com/stm32duino/Arduino_Core_STM32/assets/37047660/d8adf9cb-fb2e-4aa7-a5b1-0dca509e8b9d)

for example in my example wire.write ("Hello/n") it means 6byte are sending after that STM is not working 

From the master you can try to read more bytes instead of 6 bytes. If you try more than 6 bytes than STM freezes. 
Wire.requestFrom(slave_address,50,true);

**Reply by fpistm** [reply]
Is it a legacy STM32F103?

**Reply by gopalreddylakki** [reply]
Yes I am using BLUE PILL (STM32F103C)

**Reply by fpistm** [reply]
This does not mean it is a real STM32, several "bluepill" have a fake STM32.

**Comment by gopalreddylakki** [comment]
![IMG_20240220_150609](https://github.com/stm32duino/Arduino_Core_STM32/assets/37047660/8933e044-a920-4ff9-9d44-e818ad9b72ee)
in Our PCB also not working , in BLUE PILL SO ORGINAL ONE

**Comment by gopalreddylakki** [comment]
any update sir

**Reply by fpistm** [reply]
No. It works on my side with the default examples master/reader. So hard to help on this.

**Reply by gopalreddylakki** [reply]
when before sending string, send wire.write(0) then its not freeze

**Comment by gopalreddylakki** [comment]
![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/37047660/168e5a47-9172-4501-abab-7a0743f3c913)


**As for this write if its successfully send then return 0 , but in the program getting Len of string 6 only** 



![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/37047660/16b4e858-37f6-4f5c-9e60-fae46a11ba9b)

2- is request address 
6-is length of string

**Reply by fpistm** [reply]
> **As for this write if its successfully send then return 0 , but in the program getting Len of string 6 only**

No, it return `ret = quantity` as expected. 0 is error.

**Reply by gopalreddylakki** [reply]
### if have the board try load this code try to read on i2c more than 6 byte than STM is frizzing can plz check it , less than 6 byte working fine but i am requesting more than 6 byte is freezing 






#include <Wire.h>

#define I2C_ADDR  8

TwoWire Wire1(PB11,PB10);
void setup()
{
  Wire1.begin(I2C_ADDR);         // join i2c bus with address #4
  Wire1.onRequest(requestEvent); // register event
  Wire1.onReceive(receiveEvent); // register event
  Serial.begin(115200);           // start serial for output
}

void loop()
{
  //empty loop
}

// function that executes whenever data is received from master
// this function is registered as an event, see setup()
void receiveEvent(int howMany)
{
  while(1 < Wire1.available()) // loop through all but the last
  {
    char c = Wire1.read();     // receive byte as a character
    Serial.print(c);          // print the character
  }
  int x = Wire1.read();        // receive byte as an integer
  Serial.println(x);          // print the integer
}

// function that executes whenever data is requested by master
// this function is registered as an event, see setup()
void requestEvent()
{
 Serial.println( Wire1.print("hello\n"));  // respond with message of 6 bytes
                          // as expected by master
}

**Reply by gopalreddylakki** [reply]
Can please check it
